### PR TITLE
feat: add function verify_proof_and_register

### PR DESCRIPTION
**Pull request by hduoc2003**
# Check List

- [x] Don't forget to squash commits into meaningful chunks before merging
- [x] Check every test passed.
- [x] Did you split imports into std, third parties, and custom parts?
- [x] Format your code.
  - [x] cargo fmt
  - [x] cargo clippy
  - [x] `taplo fmt --config taplo/taplo.toml`
- [x] Check unwrap, expect().
- [x] Check clone().
- [x] Check your commit messages.
- [x] Have you added meaningful comments.

### CODE REVIEW

File: src/contracts_caller/gps/types/verify_proof_and_register.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
+use aptos_sdk::move_types::u256::U256;
+use serde::Deserialize;
+use std::str::FromStr;
```
Comment (zk-steve) [inline src/contracts_caller/gps/types/verify_proof_and_register.rs]: group imports

File: tests/flow_test.rs [hunk 2]
```diff
@@ -20,40 +22,61 @@ mod tests {
...
                 let module_account_private_key = accounts.first().unwrap();
+                println!("module_account_private_key: {}", module_account_private_key);
                 let module_account =
                     LocalAccount::from_private_key(module_account_private_key, 0).unwrap();
+                println!("module_account_address: {}", module_account.address());
 
                 let sender_account_private_key = accounts.get(1).unwrap();
 
                 let module_address = module_account.address().to_string();
+                println!("sender_account_private_key: {}", sender_account_private_key);
```
Comment (zk-steve) [inline tests/flow_test.rs]: info!

File: src/contracts_caller/gps/verify_proof_and_register.rs [hunk 3]
```diff
@@ -0,0 +1,90 @@
+use crate::config::AppConfig;
+use crate::contracts_caller::gps::types::verify_proof_and_register::{
+    VerifyProofAndRegisterData, VerifyProofAndRegisterDataJson,
+};
+use crate::contracts_caller::transaction_helper::{get_event_from_transaction, send_tx};
+use aptos_sdk::move_types::value::MoveValue;
+use aptos_sdk::rest_client::aptos_api_types::MoveType;
+use std::fs;
+use std::str::FromStr;
```
Comment (zk-steve) [inline src/contracts_caller/gps/verify_proof_and_register.rs]: .

File: src/contracts_caller/gps/verify_proof_and_register.rs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    for fact in &data.pre_registered_facts {
+        send_tx(
+            config,
+            "fact_registry",
+            "register_fact",
+            &vec![MoveValue::U256(*fact)],
+        )
+        .await?;
+    }
+    println!("prepush_task_metadata");
```
Comment (zk-steve) [inline src/contracts_caller/gps/verify_proof_and_register.rs]: debug!

File: data
Comment (zk-steve) [inline data:1]: ?

File: dcm.py
Comment (zk-steve) [inline dcm.py:1]: ?

File: src/contracts_caller/gps/verify_proof_and_register.rs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+                data.cairo_aux_input
+                    .iter()
+                    .map(|v| MoveValue::U256(*v))
+                    .collect(),
+            ),
+            MoveValue::U256(data.cairo_verifier_id),
+        ],
+    )
+    .await?;
+    let mut time = 0;
```
Comment (zk-steve) [inline src/contracts_caller/gps/verify_proof_and_register.rs]: time ?

File: tests/flow_test.rs [hunk 6]
```diff
@@ -87,10 +109,12 @@ mod tests {
                         fri_step_size,
                         expected_root,
                     )
-                    .await
-                    .unwrap();
+                    .await?;
```
Comment (zk-steve) [inline tests/flow_test.rs]: remove

File: tests/flow_test.rs [hunk 7]
```diff
@@ -87,10 +109,12 @@ mod tests {
...
                     info!("Verify FRI {} success", i);
                 }
+                verify_proof_and_register(&config, &verify_proof_and_register::sample_data(1)?)
+                    .await?;
```
Comment (zk-steve) [inline tests/flow_test.rs]: unwrap()

File: tests/flow_test.rs [hunk 8]
```diff
@@ -64,8 +87,7 @@ mod tests {
                         height,
                         expected_root,
                     )
-                    .await
-                    .unwrap();
+                    .await?;
```
Comment (zk-steve) [inline tests/flow_test.rs]: remove

File: tests/flow_test.rs [hunk 9]
```diff
@@ -19,40 +21,61 @@ mod tests {
                 let aptos_container = lazy_aptos_container().await.unwrap();
                 let node_url = aptos_container.get_node_url();
                 let module_account_private_key = accounts.first().unwrap();
+                info!("module_account_private_key: {}", module_account_private_key);
                 let module_account =
                     LocalAccount::from_private_key(module_account_private_key, 0).unwrap();
+                info!("module_account_address: {}", module_account.address());
```
Comment (zk-steve) [inline tests/flow_test.rs]: remove

File: tests/flow_test.rs [hunk 10]
```diff
@@ -19,40 +21,61 @@ mod tests {
                 let aptos_container = lazy_aptos_container().await.unwrap();
                 let node_url = aptos_container.get_node_url();
                 let module_account_private_key = accounts.first().unwrap();
+                info!("module_account_private_key: {}", module_account_private_key);
```
Comment (zk-steve) [inline tests/flow_test.rs]: remove

File: tests/flow_test.rs [hunk 11]
```diff
@@ -19,40 +21,61 @@ mod tests {
...
                 let module_account =
                     LocalAccount::from_private_key(module_account_private_key, 0).unwrap();
+                info!("module_account_address: {}", module_account.address());
 
                 let sender_account_private_key = accounts.get(1).unwrap();
 
                 let module_address = module_account.address().to_string();
+                info!("sender_account_private_key: {}", sender_account_private_key);
```
Comment (zk-steve) [inline tests/flow_test.rs]: remove

File: tests/flow_test.rs [hunk 12]
```diff
@@ -19,40 +21,61 @@ mod tests {
...
+                let sender_account =
+                    LocalAccount::from_private_key(sender_account_private_key, 0).unwrap();
+                info!("sender_account_address: {}", sender_account.address());
```
Comment (zk-steve) [inline tests/flow_test.rs]: remove

File: src/contracts_caller/transaction_helper.rs [hunk 13]
```diff
@@ -90,3 +95,33 @@ pub fn get_event_from_transaction(
...
+    let tx = build_transaction(payload, &config.account, config.chain_id);
+    let transaction = config.client.submit_and_wait(&tx).await?.into_inner();
+    let transaction_info = transaction.transaction_info()?;
+    info!(
+        "Finished running the function {} {}; gas used: {}",
+        fn_name,
+        transaction_info.hash.to_string(),
+        transaction_info.gas_used
+    );
+    if !transaction_info.success {
```
Comment (zk-steve) [inline src/contracts_caller/transaction_helper.rs]: use ensure! instead

File: src/contracts_caller/gps/verify_proof_and_register.rs [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+pub fn sample_data(test_num: isize) -> anyhow::Result<VerifyProofAndRegisterData> {
+    let data = serde_json::from_str::<VerifyProofAndRegisterDataJson>(
+        fs::read_to_string(format!(
+            "src/data_samples/gps/verify_proof_and_register_{}.json",
+            test_num
+        ))?
+        .as_str(),
+    )?;
+    Ok(VerifyProofAndRegisterData::from(data))
+}
```
Comment (zk-steve) [inline src/contracts_caller/gps/verify_proof_and_register.rs]: move this to test file

File: src/contracts_caller/gps/verify_proof_and_register.rs [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+            break;
+        }
+    }
+    Ok(())
+}
+
+// #[cfg(test)]
+pub fn sample_data(test_num: isize) -> anyhow::Result<VerifyProofAndRegisterData> {
+    let data = serde_json::from_str::<VerifyProofAndRegisterDataJson>(
+        fs::read_to_string(format!(
```
Comment (zk-steve) [inline src/contracts_caller/gps/verify_proof_and_register.rs]: include_str!()

instead

Comment (hduoc2003) [reply src/contracts_caller/gps/verify_proof_and_register.rs]: It's only work with literal string

File: Cargo.toml [hunk 16]
```diff
@@ -22,7 +22,7 @@ tokio = { version = "1.39.3", features = ["macros", "rt-multi-thread"] }
 url = { version = "2.4.0", features = ["serde"] }
 
 [dev-dependencies]
-aptos-testcontainer = { git = "https://github.com/sota-zk-labs/aptos-testcontainer.git", tag = "v0.0.3", features = [
+aptos-testcontainer = { git = "https://github.com/sota-zk-labs/aptos-testcontainer.git", tag = "v0.1.0", features = [
```
Comment (zk-steve) [inline Cargo.toml]:
```suggestion
aptos-testcontainer = { version = "0.1.2", features = [
```

**Review by Draply — APPROVED** [review]

**Pull request merged**